generator/webapp/src/main/java/io/fabric8/maven/generator/webapp/WebAppGenerator.java

```diff
@@ -0,0 +1,190 @@
+
+    private static final String DOCKER_TOMCAT_FROM = "fabric8/tomcat-8";
+    private static final String DOCKER_JETTY_FROM = "fabric8/jetty-9";
+    private static final String DOCKER_WILDFLY_FROM = "jboss/wildfly";
+
+    //TODO need to update the s2i images
+    private static final String VANNILA_S2I_FROM = "fabric8/tomcat-8";
+    private static final String REDHAT_DOCKER_S2I = "jboss-eap-6/eap64-openshift";
+
+    public WebAppGenerator(MavenGeneratorContext context) {
+        super(context, "webapp");
+    }
+
+    private enum Config implements Configs.Key {
+        deploymentDir  {{ d = "/deployments"; }},
+        user           {{ d = "jboss:jboss:jboss"; }},
+        cmd            {{ d = "/opt/tomcat/bin/deploy-and-run.sh"; }},
+        webPort        {{ d = "8080"; }},
+        jolokiaPort    {{ d = "8778"; }};
+
+        public String def() { return d; } protected String d;
+    }
+
+
+    @Override
+    public boolean isApplicable(List<ImageConfiguration> configs) {
+        return shouldAddDefaultImage(configs) &&
+                MavenUtil.hasPlugin(getProject(), "org.apache.maven.plugins:maven-war-plugin");
+    }
+
+    protected Map<String, String> getEnv() {
+        Map<String, String> defaultEnv = new HashMap<>();
+        defaultEnv.put("DEPLOY_DIR", getConfig(Config.deploymentDir));
+        return defaultEnv;
+    }
+
+    @Override
+    public List<ImageConfiguration> customize(List<ImageConfiguration> configs) {
+        ImageConfiguration.Builder imageBuilder = new ImageConfiguration.Builder();
+
+        BuildImageConfiguration.Builder buildBuilder = new BuildImageConfiguration.Builder()
+                .assembly(createAssembly())
+                .from(getFrom())
+                .ports(extractPorts())
+                .cmd(getDockerRunCommand())
+                .env(getEnv());
+        addLatestTagIfSnapshot(buildBuilder);
+        imageBuilder
+                .name(getImageName())
+                .alias(getAlias())
+                .buildConfig(buildBuilder.build());
+        configs.add(imageBuilder.build());
+
+        return configs;
+    }
+
+    protected List<String> extractPorts() {
+        List<String> answer = new ArrayList<>();
+        addPortIfValid(answer, getConfig(Config.webPort));
+        addPortIfValid(answer, getConfig(Config.jolokiaPort));
+        return answer;
+    }
+
+    private void addPortIfValid(List<String> list, String port) {
+        if (Strings.isNotBlank(port)) {
+            list.add(port);
+        }
+    }
+
+    private AssemblyConfiguration createAssembly() {
+        return new AssemblyConfiguration.Builder()
+                .basedir(getDefaultBaseDir())
+                .user(getConfig(Config.user))
+                .descriptorRef("webapp")
+                .build();
+    }
+
+
+    @Override
+    protected String getFrom() {
+        String from = super.getFrom();
+
+        if (from == null) {
+
+            if (isJetty()) {
```

Comment:
Instead of many `if (is...())` I would prefer an `AppServerDetector` interface with methods

```java
boolean isApplicable()
String getFrom()
String getCommand()
String getDeploymentDir()
```

Then you could create a list of these detectors (either in the generator or a helper class) and iterate over it until the first `isApplicable()` returns true. All the other parameters can then get from the found detector

Advantages:

* Easier to extend for future appservers (you only need to add new implementation of this interface to the list instead of finding and changing all `if(...)`
* Detection is only called once and not with every `isTomcat()`